### PR TITLE
Use PyTorch Min/Max in Inductor index propagation

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #184592
* __->__ #184589

SymPy's builtin Min/Max constructors run generic pairwise dominance checks in _find_localzeros via _is_connected. For dynamic avg_pool3d_backward CPU codegen, index propagation builds many boundary expressions containing Max(0, x - 1) and Min(s, Max(...) + k). During CPU vector tiling analysis those expressions are repeatedly substituted and simplified, and the builtin SymPy Min/Max path dominates compile time.

Use torch.utils._sympy.functions.Min/Max for symbolic ops.minimum/maximum propagation, and for avg-pool divisor boundary construction. The PyTorch Min/Max implementation keeps the useful numeric 0/1 special cases but avoids the generic pairwise is-connected test.

Benchmarks on test/inductor/test_torchinductor_codegen_dynamic_shapes.py::DynamicShapesCodegenCpuTests::test_avg_pool3d_backward2_dynamic_shapes_cpu:

- Baseline OSS repro: 1 passed in 342.93s (pytest total 0:05:42).

- Fresh-cache timed reproduction: test body 337.347s, total 352.308s.

- Compile-time breakdown before: compile_inner 333.266s, GraphLowering.codegen 330.383s, Scheduler.codegen 313.728s, compile_file 1.583s.

- SymPy instrumentation before: Min._find_localzeros 140.7s total, Max._find_localzeros 125.6s total, with ~294k Max(2 args) and ~162k Min(2 args) calls.

- Global/custom Min/Max experiment: ~72s.

- After this patch: 1 passed in 75.52s (0:01:15).

Additional validation:

- python -m pytest test/inductor/test_torchinductor.py -k 'index_propagation and cpu' -q -s: 7 passed, 2558 deselected in 16.85s.

- python -m pytest test/inductor/test_torchinductor_codegen_dynamic_shapes.py -k 'avg_pool3d and dynamic_shapes_cpu' -q -s: 3 passed, 2370 deselected, 1 xfailed in 80.10s.

Authored by Codex

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/index_propagation.py [hunk 1]
```diff
@@ -172,12 +172,12 @@ def remainder(x: TypedExpr, y: TypedExpr) -> TypedExpr | None:
     @staticmethod
     def minimum(x: TypedExpr, y: TypedExpr) -> TypedExpr:
         result_type = torch.promote_types(x.dtype, y.dtype)
-        return TypedExpr(sympy.Min(x.expr, y.expr), result_type)
```
Comment (eellison) [inline torch/_inductor/index_propagation.py:175]: Is this something we can lint away ?

**Review by eellison — APPROVED** [review]

File: torch/_inductor/codegen/pallas.py [hunk 2]
```diff
@@ -2392,7 +2396,7 @@ def _build_load_expr(
             self.has_flatten_indexing = True
             self.flatten_indexed_buffers.add(name)
             # Flatten then index for non-contiguous access (gather operation)
-            has_minmax = index.has(sympy.Min) or index.has(sympy.Max)
+            has_minmax = index.has(sympy.Min, sympy.Max, TorchMin, TorchMax)
```
Comment (ezyang) [inline torch/_inductor/codegen/pallas.py:2399]: Mergedog change endorsed, but I am now worried I haven't hit all the sympy.Max sites now...

**Review by laithsakka — APPROVED** [review]

**Comment by ezyang** [comment]
## mergedog handoff

All CI is green (or skipped). Ready for human review and `@pytorchbot merge`.

Current PR head: `dd131ede25f3dfaa4a977f34bdb3e7ef9556442d`.

### Autonomous changes pushed

- `128228dbe7e1` — pushed fix commit 128228dbe7e1 — failing jobs: macos-py3-arm64 / test (mps, 1, 1, macos-m2-15) (fix-CI (2026-05-21T04:18:17+00:00))

During shepherding, the configured LLM was invoked **1** time. If any CI shows red, it judged that failure unrelated to this PR's changes — please verify below before merging.

### Session 1 — fix-CI (2026-05-21T04:18:17+00:00)

- **Before:** `dd131ede25f3`
- **After:** `128228dbe7e1`
- **Verdict:** pushed fix commit 128228dbe7e1 — failing jobs: macos-py3-arm64 / test (mps, 1, 1, macos-m2-15)

<details><summary>LLM transcript</summary>

```
```

</details>

File: torch/_inductor/codegen/mps.py [hunk 3]
```diff
@@ -93,23 +93,25 @@ def _print_ModularIndexing(self, expr: sympy.Expr) -> str:
...
 
     def _print_Max(self, expr: sympy.Expr) -> str:
-        if len(expr.args) != 2:
-            raise RuntimeError("metal::max only supported for 2 args")
-        # pyrefly: ignore [missing-attribute]
-        a, b = map(self._print, expr.args)
-        typecast_a = f"static_cast<decltype({a}+{b})>({a})"
-        typecast_b = f"static_cast<decltype({a}+{b})>({b})"
-        return f"metal::max({typecast_a}, {typecast_b})"
+        return self._print_min_max(expr, "max")
```
Comment (ezyang) [inline torch/_inductor/codegen/mps.py:114]: Oh wow! Big win from mergedog
